File: service/src/main/java/no/nav/melosys/service/kontroll/feature/ferdigbehandling/kontroll/FerdigbehandlingKontrollsett.java
package no.nav.melosys.service.kontroll.feature.ferdigbehandling.kontroll;

import java.util.Set;
import java.util.function.Function;

import no.nav.melosys.domain.kodeverk.Sakstyper;
import no.nav.melosys.service.kontroll.feature.ferdigbehandling.data.FerdigbehandlingKontrollData;
import no.nav.melosys.service.validering.Kontrollfeil;

public class FerdigbehandlingKontrollsett {

    public static Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> hentRegelsettForVedtak(Sakstyper sakstype,
                                                                                                   boolean harRegistreringUnntakFraMedlemskapFlyt,
                                                                                                   boolean harIkkeYrkesaktivFlyt) {
        if (harRegistreringUnntakFraMedlemskapFlyt) {
            return REGELSETT_UNNTAKSREGISTRERING;
        }
        if (harIkkeYrkesaktivFlyt) {
            return REGELSETT_IKKE_YRKESAKTIV;
        }
        return switch (sakstype) {
            case EU_EOS -> REGELSETT_EU_EOS;
            case FTRL -> REGELSETT_FTRL;
            case TRYGDEAVTALE -> REGELSETT_TRYGDEAVTALER;
        };
    }

    public static Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> hentRegelsettForAvslagOgHenleggelse() {
        return REGELSETT_AVSLAG_HENLEGGELSE;
    }

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_EU_EOS = Set.of(
        FerdigbehandlingKontroll::adresseRegistrert,
        FerdigbehandlingKontroll::overlappendePeriode,
        FerdigbehandlingKontroll::periodeOver24Mnd,
        FerdigbehandlingKontroll::periodeManglerSluttdato,
        FerdigbehandlingKontroll::arbeidsstedLandManglerFelter,
        FerdigbehandlingKontroll::arbeidsstedMaritimtManglerFelter,
        FerdigbehandlingKontroll::arbeidsstedOffshoreManglerFelter,
        FerdigbehandlingKontroll::arbeidsstedLuftfartManglerFelter,
        FerdigbehandlingKontroll::foretakUtlandManglerFelter,
        FerdigbehandlingKontroll::selvstendigUtlandManglerFelter,
        FerdigbehandlingKontroll::orgnrErOpphørt,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_FTRL = Set.of(
        FerdigbehandlingKontroll::adresseRegistrert,
        FerdigbehandlingKontroll::overlappendePeriode,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_TRYGDEAVTALER = Set.of(
        FerdigbehandlingKontroll::adresseRegistrert,
        FerdigbehandlingKontroll::overlappendePeriode,
        FerdigbehandlingKontroll::periodeOver12Måneder,
        FerdigbehandlingKontroll::periodeOverTreÅr,
        FerdigbehandlingKontroll::periodeOverFemÅr,
        FerdigbehandlingKontroll::periodeManglerSluttdato,
        FerdigbehandlingKontroll::arbeidsstedLandManglerFelter,
        FerdigbehandlingKontroll::representantIUtlandetMangler,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_UNNTAKSREGISTRERING = Set.of(
        FerdigbehandlingKontroll::overlappendeMedlemskapsperiode,
        FerdigbehandlingKontroll::periodeManglerSluttdato,
        FerdigbehandlingKontroll::overlappendeUnntaksperiode,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_IKKE_YRKESAKTIV = Set.of(
        FerdigbehandlingKontroll::adresseRegistrert,
        FerdigbehandlingKontroll::overlappendePeriode,
        FerdigbehandlingKontroll::periodeManglerSluttdato,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );

    private static final Set<Function<FerdigbehandlingKontrollData, Kontrollfeil>> REGELSETT_AVSLAG_HENLEGGELSE = Set.of(
        FerdigbehandlingKontroll::adresseRegistrert,
        FerdigbehandlingKontroll::åpentUtkastFinnes
    );
}

File: service/src/main/kotlin/no/nav/melosys/service/kontroll/feature/ferdigbehandling/kontroll/FerdigbehandlingKontrollsett.kt
package no.nav.melosys.service.kontroll.feature.ferdigbehandling.kontroll

import no.nav.melosys.domain.kodeverk.Sakstyper
import no.nav.melosys.service.kontroll.feature.ferdigbehandling.data.FerdigbehandlingKontrollData
import no.nav.melosys.service.validering.Kontrollfeil
import java.util.function.Function

typealias Regelsett = Set<Function<FerdigbehandlingKontrollData, Kontrollfeil?>>

object FerdigbehandlingKontrollsett {

    fun hentRegelsettForVedtak(sakstype: Sakstyper, harRegistreringUnntakFraMedlemskapFlyt: Boolean, harIkkeYrkesaktivFlyt: Boolean): Regelsett {
        if (harRegistreringUnntakFraMedlemskapFlyt) {
            return REGELSETT_UNNTAKSREGISTRERING
        }
        if (harIkkeYrkesaktivFlyt) {
            return REGELSETT_IKKE_YRKESAKTIV
        }

        return when (sakstype) {
            Sakstyper.EU_EOS -> REGELSETT_EU_EOS
            Sakstyper.FTRL -> REGELSETT_FTRL
            Sakstyper.TRYGDEAVTALE -> REGELSETT_TRYGDEAVTALER
        }
    }

    fun hentRegelsettForAvslagOgHenleggelse(): Regelsett {
        return REGELSETT_AVSLAG_HENLEGGELSE
    }

    private val REGELSETT_EU_EOS: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.adresseRegistrert(it) },
        Function { FerdigbehandlingKontroll.overlappendePeriode(it) },
        Function { FerdigbehandlingKontroll.periodeOver24Mnd(it) },
        Function { FerdigbehandlingKontroll.periodeManglerSluttdato(it) },
        Function { FerdigbehandlingKontroll.arbeidsstedLandManglerFelter(it) },
        Function { FerdigbehandlingKontroll.arbeidsstedMaritimtManglerFelter(it) },
        Function { FerdigbehandlingKontroll.arbeidsstedOffshoreManglerFelter(it) },
        Function { FerdigbehandlingKontroll.arbeidsstedLuftfartManglerFelter(it) },
        Function { FerdigbehandlingKontroll.foretakUtlandManglerFelter(it) },
        Function { FerdigbehandlingKontroll.selvstendigUtlandManglerFelter(it) },
        Function { FerdigbehandlingKontroll.orgnrErOpphørt(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) },
        Function { FerdigbehandlingKontroll.storbritanniaKonvensjonBruktForTidlig(it) },
        Function { FerdigbehandlingKontroll.kunEnAvklartVirksomhet(it) })

    private val REGELSETT_FTRL: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.adresseRegistrert(it) },
        Function { FerdigbehandlingKontroll.overlappendePeriode(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) },
        Function { FerdigbehandlingKontroll.harOverlappendePeriodeMedForskuddsvisFaktureringIAnnenSak(it) },
        Function { FerdigbehandlingKontroll.direkteForutgåendePeriode(it) })

    private val REGELSETT_TRYGDEAVTALER: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.adresseRegistrert(it) },
        Function { FerdigbehandlingKontroll.overlappendePeriode(it) },
        Function { FerdigbehandlingKontroll.periodeOver12Måneder(it) },
        Function { FerdigbehandlingKontroll.periodeOverTreÅr(it) },
        Function { FerdigbehandlingKontroll.periodeOverFemÅr(it) },
        Function { FerdigbehandlingKontroll.periodeManglerSluttdato(it) },
        Function { FerdigbehandlingKontroll.arbeidsstedLandManglerFelter(it) },
        Function { FerdigbehandlingKontroll.representantIUtlandetMangler(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) })

    private val REGELSETT_UNNTAKSREGISTRERING: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.overlappendeMedlemskapsperiode(it) },
        Function { FerdigbehandlingKontroll.periodeManglerSluttdato(it) },
        Function { FerdigbehandlingKontroll.overlappendeUnntaksperiode(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) })

    private val REGELSETT_IKKE_YRKESAKTIV: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.adresseRegistrert(it) },
        Function { FerdigbehandlingKontroll.overlappendePeriode(it) },
        Function { FerdigbehandlingKontroll.periodeManglerSluttdato(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) })

    private val REGELSETT_AVSLAG_HENLEGGELSE: Regelsett = setOf(
        Function { FerdigbehandlingKontroll.adresseRegistrert(it) },
        Function { FerdigbehandlingKontroll.åpentUtkastFinnes(it) })
}
